### PR TITLE
[fmha_fwd][rocm] Extend CK ck_tile FMHA seqlen-pad guard to the GQA swap path

### PR DESCRIPTION
**Pull request by jnwan**
Summary:
Follow-up to the gfx9 CK ck_tile FMHA tile-tail OOB guard: the guard excluded
the seqlenq_ngroups_swapped (GQA decode) path, but that path is exactly where a
real MoE/MTML inference model still faults during AOTT GPU lowering on AMD
(MI300X gfx942): "Memory access fault by GPU node-N ... Reason: Unknown".

Root cause: for GQA decode (seqlen_q==1, num_heads > num_heads_k) F.sdpa's CK
flash path transposes (b,1,(hk*ngroups),d) -> (b,ngroups,hk,d), making
seqlen_q=ngroups. For the faulting request ngroups=2000 (pad 2048) and
seqlen_k=656 (pad 768) are both tile-unaligned, so CK selects a no-seqlen-pad
fwd instance that writes round_up(seqlen_q, kM0=128)=2048 rows into the
[b, ngroups, hk, d] output -- over-writing O/LSE ~0.5 MB past the buffer and
faulting (decoded via a temporary TORCH_WARN at the call site: b=32 hq=2 hk=2
sq=2000 sk=656 d=128 swap=1, last kernel before the fault =
ck_tile::FmhaFwdKernel<... TileFmhaTraits<kPadSeqLenQ=1,kPadSeqLenK=1...>>,
grid {1,2,32}). The original guard's `!seqlenq_ngroups_swapped` exclusion left
this case unprotected.

Fix:
- Drop the `!seqlenq_ngroups_swapped` exclusion so the guard engages on any
  tile-unaligned seqlen, including the swap path.
- Allocate the O/LSE scratch from the logical post-swap kernel dims
  ({batch, seqlen_q_pad, num_heads, head_size_8x}) instead of out.sizes(): on
  the swap path `out` is still [b, 1, num_heads*seqlen_q, d], so out.size(2)/(3)
  would be wrong.
- Copy the real rows back via reshape(out.sizes()) / reshape(softmax_lse.sizes())
  so the padded scratch maps onto the caller's actual layout for both the swap
  and non-swap cases.

Durable follow-up (unchanged): CK should generate/select kPadSeqLen fwd
instances upstream, after which this host guard can be removed.

Test Plan:
Reproduced and fixed on MI300X (gfx942) via the from-source lowering binary
(re_mvai_cinder) on the failing autotune request (model 2118627168), classic mode.

BEFORE (this commit absent): "Memory access fault by GPU node-2 ... Reason:
Unknown" at the scripted-model validation forward pass. With the caching
allocator disabled (PYTORCH_NO_HIP_MEMORY_CACHING=1) the same fault is
deterministic and resolves to "Write access to a read-only page" on the CK
FmhaFwd kernel, swap=1 guard=0.

AFTER (this commit):
- PYTORCH_NO_HIP_MEMORY_CACHING=1 (deterministic repro): swap-path calls now log
  guard=1, 0 memory faults, "Scripted model validation forward pass completed
  successfully", Done to_torch_package.
- Normal caching allocator (production-like): 0 memory faults, validation
  completed, "Upload lowered model ... size: 30483746255 bytes", lowering
  completed in ~402s.

Numerics: the lowering's own scripted-model validation forward pass (5 real
dumped requests) is a correctness check and passes; the swapped copy-back layout
is byte-identical to the guard-off path (same contiguous logical strides).

Differential Revision: D108947199

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @jataylo @hongxiayang @naromero77amd @pragupta @jerrymannil @xinyazhang

### CODE REVIEW

**Comment by jnwan** [comment]
@jeffdaily could you help review this PR? this PR is to extend the fix of https://github.com/pytorch/pytorch/pull/187152.

**Comment by jeffdaily** [comment]
Putting this on hold pending a standalone CK reproducer rather than merging it.

The description contradicts itself on the root cause: it says CK "selects a no-seqlen-pad fwd instance", then reports the decoded kernel as `TileFmhaTraits<kPadSeqLenQ=1, kPadSeqLenK=1, ...>`. Those cannot both be true, and the difference matters. In the CK pin we carry (`5a74dec07`), the non-pad instances have real dispatch guards (`scheck` is `a.seqlen_q % {bm0} == 0`, `skcheck` similarly on `bn0`), while the pad-enabled instance's guard is hardcoded `true` with an `order of get_pipelines() matters (ugly)` TODO. If the kernel that actually faults is the pad-enabled one, then CK's predication is broken and the host-side padding is masking a kernel bug rather than working around a dispatch bug, which changes where the fix belongs.

CK's own `tile_example_fmha_fwd` takes these shapes directly and `-kname=1` prints the selected instance, so this is answerable without PyTorch in the loop. I would rather hand the CK team that plus a canary-based demonstration of the overwrite than extend the stopgap further.

Two other things regardless of how that lands. This conflicts with #189730, which gated `ck_oob_guard` to gfx950 on the premise that gfx942 does not fault; your gfx942 repro is a counterexample, so the arch gating has to be reconciled deliberately rather than by rebase. And on the GQA decode path `seqlen_k` grows by one per step, so the guard would engage on roughly 255 of every 256 calls and pad-copy K and V each time, which is the same overhead class that forced the gfx950 gate in the first place; that needs a decode-throughput number before it lands.

Comment drafted with AI assistance.

**Review by jeffdaily — CHANGES_REQUESTED** [review]
Requesting changes. I would rather not merge this as written, and I am leaving it open rather than closing it so you can decide what to do with it.

Since my hold comment I built the standalone CK reproducer I asked for and ran it on MI300X (gfx942) and MI355X (gfx950). At the CK pin we carry (`third_party/composable_kernel` at `5a74dec07`) I could not produce an out-of-bounds access in the ck_tile FMHA forward path on either architecture, including on your exact shape and on the swapped GQA layout. Evidence below. If I am missing the case, the "What would change my mind" section says what would show it.

## What I ran

A PyTorch-free program that calls `fmha_fwd(traits, args, stream_config)`, the same entry point `mha_fwd_ck.hip` uses, with PyTorch-identical strides for both the plain and the `seqlenq_ngroups_swapped` transposed-view layouts, under two detectors:

- Canary: every tensor gets a 1 MiB head and tail margin filled with 0xA5. Any modified margin byte is a proven out-of-bounds write, and its offset says exactly how far past the end the kernel wrote.
- Electric fence: the tensor under test is backed by `hipMemCreate` / `hipMemMap` / `hipMemSetAccess` so that its logical last byte is the last addressable byte and the next page is reserved but unmapped. An over-read or over-write then faults deterministically instead of depending on where the allocator happened to end the buffer. A self-test that writes one byte into the guard region confirms the fence is armed:

```
Memory Fault Error [... faulting addr: 0x762d923bb000, kernel: poke_guard(char*)]
```

That distinction matters for the arch-gating argument generally: whether an out-of-bounds access faults depends on where the allocation ends, which is an allocator property rather than an architecture property. "Does not fault on gfx942" was never evidence of safety, and the fence removes that dependence in both directions.

I linked the same program two ways: against PyTorch's own forward instance set (`--api fwd --optdim=32,64,128,256 --receipt 4 --filter "*_lse*ntrload*nsink*"`, the kernels `mha_fwd_ck.hip` can actually dispatch to) and against CK's full set from `example/ck_tile/01_fmha`, which is a superset of ours.

## Results

52 configurations per instance set on gfx950, 48 of which run the kernel: your shape (b=32, h=2, h_k=2, seqlen_q=2000, seqlen_k=656, d=128) in both the plain and the swapped layout, the #187152 shape (b=5, h=2, seqlen_q=4, seqlen_k=656, d=128), an aligned control, a ten point unaligned sweep at b=1 so that each tensor's tail is the buffer tail, d=64 and d=256, and a GQA shape with h=8 and h_k=2. Each shape ran with the fence on q, k, v, o and lse individually and on all five at once, plus a canary run validated against a CPU reference.

| | PyTorch instance set | CK instance set |
|---|---|---|
| kernel runs | 48 | 48 |
| canary violations | 0 | 0 |
| memory access faults | 0 | 0 |
| numerics mismatches vs CPU reference | 0 | 0 |

An earlier run of the same artifact on gfx942, same ROCm and same CK pin, gave the same result.

## Why nothing goes out of bounds

`include/ck_tile/ops/fmha/kernel/fmha_fwd_kernel.hpp` builds every global view from the logical extent:

```cpp
const auto q_dram_naive = make_naive_tensor_view<address_space_enum::global>(
    q_ptr, make_tuple(kargs.seqlen_q, kargs.hdim_q), make_tuple(kargs.stride_q, 1), ...);
```

and `make_naive_tensor_view<address_space_enum::global>` lowers to `make_wave_buffer_resource(ptr, element_space_size * sizeof(T))` (`include/ck_tile/core/arch/amd_buffer_addressing.hpp`), an AMD buffer resource whose `num_records` is the real tensor size. Buffer loads past `num_records` return zero and buffer stores past it are discarded, in hardware. The `kPadSeqLen*` traits control the logical masking of the tile window, that is whether the computed values are correct, not memory safety. A tile-tail access cannot leave the allocation whatever the kPad flags say, which is exactly what both detectors measured.

## What the dispatcher can actually select

Parsing the checked-in `fmha_fwd_api.hip` (receipt 4) for gfx950 gives 272 candidates: 192 with `kPadS=true, kPadSK=true`; 32 with `kPadS=true, kPadSK=false`, each guarded by `a.seqlen_k % kN0 == 0`; and 48 with `kPadS=false`, each guarded by `a.seqlen_q % kM0 == 0`. Every no-pad candidate carries a real alignment guard, so a tile-unaligned seqlen cannot select one. For bf16, no bias, no mask, lse, no dropout, d=128 there are exactly four candidates and all four are `kPadS=true`; `seqlen_k=656` fails `% 128 == 0`, so the `kPadSK=true` one is selected. That is what the probe dispatches, and it agrees with the trait string you decoded (`kPadSeqLenQ=1, kPadSeqLenK=1`). So the summary's "CK selects a no-seqlen-pad fwd instance" is not what happens on this pin, and the pad-enabled instance that does run is not writing `round_up(2000, 128) = 2048` rows into a 2000-row output: I fenced that output, on your shape, in both layouts, and it did not fault.

## The one place gfx950 differs, and it still is not a bug

On gfx950, CK's own instance set does dispatch a no-pad instance for both shapes, `fmha_fwd_d128_bf16_batch_b16x32x64x128x32x128_..._qr_async_trload_vr_npad_...`, meaning `kPadSeqLenQ = kPadSeqLenK = false` with kM0=16 and kN0=32 against seqlen_q=4 and seqlen_k=656. That is literally the situation #187152 describes. It still does not access out of bounds under the fence, in 33 of the 48 CK-set runs, and CK's own `-kname=1 -v=1` run reports `valid:y` for it. PyTorch never sees it in any case: the `*ntrload*` part of our codegen filter excludes the whole trload family.

## Cost

Measured on gfx950 with the CK backend forced, since that is the architecture the guard is currently gated to:

```
#187152 shape, b=5 h=2 s_q=4 s_k=656 d=128:         pad+copy 0.094 ms vs 0.039 ms padded compute
prefill, b=4 h=8 s_q=1000 s_k=1000 d=128:           pad+copy 0.127 ms vs 0.034 ms padded compute
decode, b=32 h_q=16 h_k=2 s_q=1 s_k=656:            K+V pad  0.049 ms vs 0.048 ms decode step
```

The last line is the path this PR targets, and it is roughly a doubling per step. It is also the common case rather than the exception: seqlen_k grows by one per decode step, so it is tile-unaligned on 255 of every 256 steps. That measured pad cost is a lower bound, since it counts K and V only and not the O and LSE pad and copy-back that an unaligned ngroups adds.

## Merge state

This is currently CONFLICTING against main. The branch predates #189730, so its base has no `isGPUArch({"gfx950"})` in `ck_oob_guard` at all. A rebase has to make a deliberate decision about the arch gate rather than inherit one.

## What would change my mind

The reproducer is standalone, so if the fault is real on your configuration this should be quick to pin down. I can share the source if you want to run it against your build.

1. The exact torch commit, ROCm version and `third_party/composable_kernel` pin the fault was seen on. If the CK pin predates `5a74dec07`, this may be a real bug that CK has since fixed, which is a different conclusion worth recording rather than a workaround worth extending.
2. Whether the fault reproduces through eager `F.scaled_dot_product_attention` on that shape, rather than only inside the lowering pipeline. This PR and the #187152 guard comment both attribute their faults to "AOTT lowering", and I do not know that path. If the faulting access comes from somewhere other than `fmha_fwd`, the guard is in the wrong place regardless of the arch gate.
3. The details behind "Write access to a read-only page" under `PYTORCH_NO_HIP_MEMORY_CACHING=1`: the faulting address, the O buffer base address, and the kernel name. An offset of 48 rows past the end of O would confirm the stated mechanism; an address unrelated to O points elsewhere. The read-only detail is itself worth explaining, since an overrun off the end of an O allocation would normally land on unmapped or read-write memory rather than a read-only page.

## Question on the original fault

@goldcoderZ you wrote both #187152 and #189730, so you have context I do not. What was the original MI350X fault: which CK pin, was it ever reproduced outside AOTT lowering, and was there an address-level or canary demonstration of the O/LSE overwrite as opposed to a fault that the padding happened to make go away? If it was only ever seen through that lowering path, then on the current pin the gfx950 guard is costing roughly its own kernel again on the shapes where it engages while protecting against something the hardware buffer bounds checking already handles, and the follow-up is to remove it rather than to extend it.

Review drafted with AI assistance.

**Comment by jeffdaily** [comment]
Reproducer is here, since it is easier to argue against something runnable: https://gist.github.com/jeffdaily/d768e19c4c9f3672ae355087bce03b4b

`fmha_oob_repro.cpp` is the probe itself, `build_repro.sh` links it against PyTorch's forward instance set and `build_repro_ckset.sh` against CK's full set, `run_all.sh` is the configuration matrix, and `results_pt_set.log` / `results_ck_set.log` are the raw output behind the numbers in my review. `bench_guard_cost.py` is the cost measurement. It takes `$ROCM`, `$PYTORCH_ROOT` and `$TORCH_INC` from the environment, and the only slow part is building the CK instances.

If it does fault on your configuration, that output is exactly what I am missing, and I would rather see it than keep guessing at the mechanism from here.

Comment drafted with AI assistance.
